### DISCUSSION TITLE
How do you share state between load functions

### DISCUSSION THREAD

**Discussion by AndreasHald** [sveltejs/kit#10715] [Q&A]
Hi, I'm using a graphql client in svelte kit particularly I'm using Apollo-client but I believe the issue applies to all other clients swell.

The concept of the graphql-client is that all graphql requests goes through it and I can do a lot of caching and state management.

The general idea is as follows

```ts
/* routes/+layout.ts */

export const load: LayoutLoad = async (page) => {
    const graphqlClient = buildClient(page.fetch)

   return {
      graphqlClient
   }
});

```

```svelte
<!-- routes/+layout.svelte-->

<script lang="ts">
   export let data;
   setContext('client', data.graphqlClient);
</script>

<slot />
```

And now all subsequent components can access the graphql client on the context and use it to query data.

The problem is that if I wish to use it in a subsequent load function I have to await parent data potentially causing waterfalls

```ts
/* routes/foo/bar/+page.ts */

export const load: LayoutLoad = async (page) => {
   const { client } = await page.parent();

   const someData = await client.query(...);
   return {
      someData
   }
});
```

Does anyone have a good idea on how to avoid this? while preserving the fact that I can use a single client for all graphql requests?

**Comment by david-plugge** [comment]
There isn't a clean way to achieve this at the moment. I tried to convince the team multiple times that we absolutely need this functionality.
Like a `handle`Hook for shared load functions

**Comment by stephane-vanraes** [comment]
You can use the `handle` hook to put your client on `locals`.

```js
// hooks.server.js
export const handle = async ({ event, resolve }) => {
  event.client = buildGraphQlClient();
  return await resolve(event);
}
```

```js
// /any_other_page/+page.server.js
export const load = async ({ locals }) => {
   const data = await locals.client.get('stuff_stuff');
}
```

**Reply by david-plugge** [reply]
That only works for server load functions

**Reply by stephane-vanraes** [reply]
Sorry misread the question, if you want this shared across the _client_, you can simply move it to an external file and import it from there wherever you need it 

```js
const client = buildGraphQlClient();
export default client;
```

this code will only run the first time you actually import it as well, so you will not be making the client if you don't need it.

One thing to not forget is that `load` in a `+page.ts` also runs on the server for the first request, where using stores is not really recommended, but you can combine both: 

```js
import { browser } from '$app/environment';
import { client } from '$lib/graphql';

export const load = ({ locals }) => {
 const my_client = browser ? client : locals.client;
}
```

Now you have 1 client for the server and 1 for the browser (you would have to have two different ones here anyway)

**Reply by david-plugge** [reply]
And how do you provide `event.fetch` initially when creating the client without doing it in every load function? That's what we are interested in.

**Reply by stephane-vanraes** [reply]
In your handle hook you pass `event.fetch` to the builder

```js
// hooks.server.js
export const handle = async ({ event, resolve }) => {
  event.client = buildGraphQlClient(event.fetch);
  return await resolve(event);
}
```

in the `browser` branch of the load function (and anywhere else you are on the client) you use the client and have that one use the native fetch

```js
import { browser } from '$app/environment';
const client = browser ? buildGraphQlClient(fetch) : () => void 0;
export default client;
```

since this file will only run only on the client you can simply use the global fetch here (not sure the browser check is required, but it probably doesn't harm)

**Reply by david-plugge** [reply]
I want the client to use `event.fetch` in shared load functions aswell. The handle hook only runs on the server. In the example you gave the graphql client inside
- handle hook and server load functions would use `event.fetch`
- shared load functions would always use the global fetch (no difference in not specifiying fetch and doing so, it takes the fetch from `window` or `globalThis` on both cases)

As you can see, there is just no way to do this without calling `await parent()` and therefore introducing a waterfall.

**Reply by david-plugge** [reply]
and setting the client to `() => void 0` would just end up in an error on the server?

**Reply by AndreasHald** [reply]
Thanks for chiming in @stephane-vanraes :) 

It's an interesting approach, but I don't think it could work. the goal is to be able to use the graphql client in +page.js files, and they do not have access to locals - so I would need to move all the logic into +page.server.js files, however that defeats the purpose of graphql, since I would send requests to each route specifically instead of a singular graphql endpoint.

**Comment by stephane-vanraes** [comment]
Starting a new thread because the other one went a bit astray.
So, a main issue is that you want a "shared client" that you can also pass the `event.fetch` too.

Let's say we make a file 
```js
// lib/graphql/client.js
let client;

export default function getClient(fetch) {
  if (client) return client;
  client = buildClient(fetch);
  return client;
}
```

Now using this function in a `load` will always return the same instance once it has been called for the first time.
In the cases where `load` function is executed server-side (first render) it will use the sveltekit fetch instead of the regular fetch and it will still work without any change.

Obviously there is _no way_ to share the graphQlClient between client/server, those will be two different instances.

**Reply by AndreasHald** [reply]
No problem, this is also one of the first solutions I attempted, and the only problem it has is that the client is now not bound to a single request on the server. Which means that if User A and User B hits the server simultaneously the graphql client will be populated with data from both users. In "open" apps without any security this would probably solve all issues, however if data is private this cannot work.

**Reply by AndreasHald** [reply]
The fix I ended up with was that I build the graphql client in the outermost +layout.js file, and then pass it to all subsequent load files and page files from there. This works great in that it isolates the graphql client to each request. the only issue is that it can introduce waterfalls depending on your route structure, because all subsequent load function has to await the parent to get the client

**Reply by AndreasHald** [reply]
An alternative approach I could think up, however ended up not doing was something like

```js
let clients: {};

export default function getClient(fetch, requestId) {
 if(clients[requestId]) return clients[requestId]
 clients[requestId] = buildClient(fetch)
 return clients[requestId]
}
```

this could scope the client, to a specific user, however it would explode the memory footprint on the server.

**Reply by david-plugge** [reply]
As @AndreasHald mentioned, that doesnt work in some cases. What we actually need is a hook to initialize the client and pass it to all shared load functions.

```ts
export const initLoad = ({event, resolve}) => {
  event.shared.client = createClient(event.fetch)
  return resolve(event)
}
```
This looks very similar to the handleHook and it should, because it would be the `shared` entry point for load functions, just like the handle hook is for all server load functions.

**Reply by stephane-vanraes** [reply]
> Which means that if User A and User B hits the server simultaneously the graphql client will be populated with data from both users.

You could solve this by checking for browser in the build function

```js
// lib/graphql/client.js
import { browser } from '$app/environment';
let client;

export default function getClient(fetch) {
 if (client) {
   return client;
 }
 else if (browser) {
   client = buildClient(fetch);
   return client;
 }
 else {
   return buildClient(fetch);
 }
}
```

Now if you are on the server you will create a new graphql client per user, while on the client you will have a shared one.
Or am I still missing something ?

**Reply by AndreasHald** [reply]
This is also a very pragmatic approach. I may actually try it out because it may be more performant than water falling in my case. However it does not solve the underlying issue.

One of the advantages of the graphql client is that if multiple components / pages send the same request it will cache them internally.

So for example you can do something like this

```js
// routes/foo/+page.js
export const load({fetch}) {
   const client = getGraphqlClient(fetch);
   const ressource = client.getRessource(...)
}
```

```svelte
<!-- routes/foo/+page.svelte -->

<script>
   import SomeComponent from '$lib/some-component.svelte'
</script>

<SomeComponent />
```

```svelte
<!-- lib/some-component.svelte -->

<script>
   const client = getContext('client');

   const someRessource = client.getRessource(...);
</script>

{JSON.stringify($someRessource)}
```

In this case ```someRessource``` can actually be server side rendered because when the load function runs, the ressource is put into the graphql client's cache, when some component when asks for the ressource it doesn't have to send a request, because it is already in the cache, it can synchronously be available.

While I know you can achieve this using svelte natively by simply passing the ressource from parent / child or using the built-in context. it does not provide the same flexibility as a singular client, that operates no matter how deep the component is.

The proposed solution above would not allow this to work, since ```someComponent``` and the load function would be given 2 separate instances of the client.

**Reply by david-plugge** [reply]
@AndreasHald exactly. To achieve this we need shared state between shared load functions, just as the title says. `event.params` can be used as a workaround as it´s the only thing that is the same instance/reference for all load functions. You can attach your client to that after creation and check if it exists before creating a new one. I´ve tried things like 
```ts

// lib/client.ts
const clientSymbol = new Symbol();

export function getClient(event: LoadEvent) {
  return (event.params[clientSymbol] ??= createGraphQLClient(event.fetch));
}

// +page.ts
export const load = (event) => {

  const client = getClient(event);

  return {
    ...
  }
}
```
It works but it´s ugly.

**Reply by AndreasHald** [reply]
@david-plugge is this solution scoped per request on the server? event.params is not shared between any separate requests?

**Reply by david-plugge** [reply]
Yes. The params object is created on every request and the exact same reference is passed to each load function.
I didn't really use it and this behavior may have changed, but it's easy to test. I can provide an example when I'm at my pc.

**Reply by david-plugge** [reply]
@AndreasHald this works pretty good

```ts
import type { LoadEvent } from '@sveltejs/kit';

// use this as a factory to create "states"
export function createSharedState<T>(factory: (event: LoadEvent) => T): (event: LoadEvent) => T {
	const symbol = Symbol();
	return (event) => ((event.params as any)[symbol] ??= factory(event));
}

// create a "state"
export const getSharedRandom = createSharedState((event) => Math.random());


// inside shared load functions:
import { getSharedRandom } from '$lib';

export const load = (event) => {
	const random = getSharedRandom(event);

	return {
		randomInLayout: random
	};
};
```

the downside is obviously that you always need to pass `params` which is very unintuitive. Therefore i´d rather go with the full event like in my example above. A downside of this is that you can´t  directly use destructuring in your load functions.

**Reply by AndreasHald** [reply]
Ive seen worse hacks, im gonna try it out when i get a minute, and see if it solves all my problems.

@stephane-vanraes do you or anyone from the svelte team know if this is safe to implement?

**Reply by stephane-vanraes** [reply]
@AndreasHald it sounds  a bit hacky and feels similar to your own proposal of keeping track of all request that is at risk of causing a memory leak server side, but you could run some tests to see it it works.

**Comment by david-plugge** [comment]
I worked out a solution for universal load functions: https://gist.github.com/david-plugge/33cd821a4c829059f67ce18468b8dfca.
The same would work for server load functions aswell. Its definately a hack but still better than a waterfall. This should be built into sveltekit.
